cartographer/mapping_2d/pose_graph.cc

```diff
@@ -369,8 +385,22 @@ void PoseGraph::WaitForAllComputations() {
 }
 
 void PoseGraph::FinishTrajectory(const int trajectory_id) {
-  // TODO(jihoonl): Add a logic to notify trimmers to finish the given
-  // trajectory.
+  AddWorkItem([this, trajectory_id]() REQUIRES(mutex_) {
+    const auto& submap_data = optimization_problem_.submap_data();
+    for (auto submap_id_data : submap_data.trajectory(trajectory_id)) {
+      submap_data_.at(submap_id_data.id).state = SubmapState::kFinished;
+    }
+
+    for (auto& trimmer : trimmers_) {
+      trimmer->FinishTrajectory(trajectory_id);
+    }
+  });
+
+  // To call Trim() when there is no active trajectory
```

Comment:
This does not work. It would only call `Trim()` if there is nothing queued already. This logic must be handled inside the work item.

Reply:
As @wohe mentioned getting the concurrency right is a little tricky. Let's move that to another PR and just ignore for now that Trim() is not invoked.

Reply:
Without this change, cartographer doesn't clean up the submaps if it was the last active trajectory because no one calls `HandleWorkQueue()`. How about calling it inside of the work item?

Reply:
The situation is similar as to `PoseGraph::ComputeConstraintsForNode` and in principle you want the same logic here it you find in mapping_2d/pose_graph.cc lines lines 274 - 280. So maybe it makes sense to pull that out into a function and call it both from here and `ComputeConstraintsForNode`. But I would suggest to leave this as a TODO for this PR and solve it in a follow-up. Or write the PR first that pulls out lines 274 - 280 into it's own function, we merge it and then you call that function here.

Reply:
Ok. I finish this first and will make a quick follow-up